### ISSUE TITLE
Signal Proxying

### ISSUE THREAD

**Issue by jippi** [hashicorp/envconsul#90]
Hi!

It seem that sending a kill signal (e.g. TERM) to a envconsul instance, does not relay it to the process "under management" - the ideal solution for me would be that any SIG* envconsul gets would be proxied down to the sub-process

e.g. using envconsul under supervisor, supervisor will send a SIGTERM signal to envconsul, and the ideal solution would be that envconsul then send whatever "-kill-signal" is configured to, down to the process below before quitting on its own.

The behavior I'm currently observing is that when envconsul gets a SIGTERM, it dies and let the child-process run on forever. No passing on the signal to the child (or even reaping it)

**Comment by jippi** [comment]
Supervisor conf

```
#
# Generated by consul-template
#

[program:frontend_frontend-render_00]
command = /usr/local/bin/envconsul -splay 5s -config /etc/envconsul/config.hcl -prefix apps/frontend/env -prefix app/frontend/supervisor/frontend-render -sanitize -upcase bash -c './node_modules/.bin/babel-node src/server/server.js'
directory = /var/www/frontend/htdocs
numprocs = 1
autostart = true
autorestart = true
startretries = 10
user = www-data
startsecs = 3
redirect_stderr = true
stdout_logfile_maxbytes = 5MB
stdout_logfile_backups = 1
stderr_logfile_maxbytes = 5MB
stderr_logfile_backups = 1
stopwaitsecs = 60
```

**Comment by jippi** [comment]
pstree when supervisor start envconsul

```
  ├─supervisord,11759 /usr/local/bin/supervisord -c /etc/supervisor/supervisord.conf
  │   ├─consul-template,30484 -config /etc/consul-template/config
  │   │   ├─{consul-template},30485
  │   │   ├─{consul-template},30486
  │   │   ├─{consul-template},30487
  │   │   ├─{consul-template},30488
  │   │   ├─{consul-template},30489
  │   │   ├─{consul-template},30490
  │   │   ├─{consul-template},30491
  │   │   ├─{consul-template},30492
  │   │   ├─{consul-template},30493
  │   │   ├─{consul-template},2532
  │   │   ├─{consul-template},17325
  │   │   └─{consul-template},17327
  │   ├─envconsul,26808 -splay 5s -config /etc/envconsul/config.hcl -prefix apps/frontend/env -prefix app/frontend/supervisor/frontend-render -sanitize -upcase ...
  │   │   ├─node,26815 ./node_modules/.bin/babel-node src/server/server.js
  │   │   │   ├─node,26821 /var/www/frontend/htdocs/node_modules/babel-cli/lib/_babel-node src/server/server.js
  │   │   │   │   ├─{node},26822
  │   │   │   │   ├─{node},26823
  │   │   │   │   ├─{node},26824
  │   │   │   │   └─{node},26825
  │   │   │   ├─{node},26817
  │   │   │   ├─{node},26818
  │   │   │   ├─{node},26819
  │   │   │   └─{node},26820
  │   │   ├─{envconsul},26809
  │   │   ├─{envconsul},26810
  │   │   ├─{envconsul},26811
  │   │   ├─{envconsul},26812
  │   │   ├─{envconsul},26813
  │   │   ├─{envconsul},26814
  │   │   └─{envconsul},26816
```

pstree when envconsul gets a SIGTERM

```
  ├─node,26821 /var/www/frontend/htdocs/node_modules/babel-cli/lib/_babel-node src/server/server.js
  │   ├─{node},26822
  │   ├─{node},26823
  │   ├─{node},26824
  │   └─{node},26825
```

**Comment by jippi** [comment]
More verbose test

Command 1

```
-> ps xau|grep nodejs # command 1
www-data  1010  0.0  0.0  12040  2012 pts/0    S+   15:19   0:00 grep --color=auto nodejs
```

Command 2 (and verbose log for command 3 +)

```
-> /usr/local/bin/envconsul -log-level=debug -once -splay 5s -config /etc/envconsul/config.hcl -prefix apps/frontend/env -prefix app/frontend/supervisor/frontend-render -sanitize -upcase bash -c './node_modules/.bin/babel-node src/server/server.js'
2016/03/02 15:19:18 [INFO] envconsul v0.6.1
2016/03/02 15:19:18 [INFO] (runner) creating new runner (command: [bash -c ./node_modules/.bin/babel-node src/server/server.js], once: true)
2016/03/02 15:19:18 [DEBUG] (runner) final config (tokens suppressed):

{
  "path": "/etc/envconsul/config.hcl",
  "consul": "127.0.0.1:8500",
  "prefixes": null,
  "prefix": [
    {
      "path": "apps/frontend/env",
      "format": ""
    },
    {
      "path": "app/frontend/supervisor/frontend-render",
      "format": ""
    }
  ],
  "secret": null,
  "auth": {
    "enabled": false,
    "username": "",
    "password": ""
  },
  "vault": {
    "renew": true,
    "ssl": {
      "enabled": true,
      "verify": true
    }
  },
  "ssl": {
    "enabled": false,
    "verify": true
  },
  "syslog": {
    "enabled": false,
    "facility": "LOCAL0"
  },
  "max_stale": 1000000000,
  "retry": 5000000000,
  "sanitize": true,
  "splay": 5000000000,
  "upcase": true,
  "timeout": 5000000000,
  "wait": {
    "min": 0,
    "max": 0
  },
  "kill_signal": "SIGTERM",
  "LogLevel": "debug",
  "pristine": false
}

2016/03/02 15:19:18 [INFO] (runner) creating consul/api client
2016/03/02 15:19:18 [DEBUG] (runner) setting consul address to 127.0.0.1:8500
2016/03/02 15:19:18 [DEBUG] (runner) setting consul token
2016/03/02 15:19:18 [INFO] (runner) creating vault/api client
2016/03/02 15:19:18 [DEBUG] (runner) enabling vault SSL
2016/03/02 15:19:18 [INFO] (runner) creating Watcher
2016/03/02 15:19:18 [INFO] (runner) starting
2016/03/02 15:19:18 [INFO] (watcher) adding "storeKeyPrefix(apps/frontend/env)"
2016/03/02 15:19:18 [DEBUG] (watcher) "storeKeyPrefix(apps/frontend/env)" starting
2016/03/02 15:19:18 [INFO] (watcher) adding "storeKeyPrefix(app/frontend/supervisor/frontend-render)"
2016/03/02 15:19:18 [DEBUG] (watcher) "storeKeyPrefix(app/frontend/supervisor/frontend-render)" starting
2016/03/02 15:19:18 [DEBUG] (view) "storeKeyPrefix(app/frontend/supervisor/frontend-render)" starting fetch
2016/03/02 15:19:18 [DEBUG] ("storeKeyPrefix(app/frontend/supervisor/frontend-render)") querying Consul with &{Datacenter: AllowStale:true RequireConsistent:false WaitIndex:0 WaitTime:1m0s Token: Near:}
2016/03/02 15:19:18 [DEBUG] (view) "storeKeyPrefix(apps/frontend/env)" starting fetch
2016/03/02 15:19:18 [DEBUG] ("storeKeyPrefix(apps/frontend/env)") querying Consul with &{Datacenter: AllowStale:true RequireConsistent:false WaitIndex:0 WaitTime:1m0s Token: Near:}
2016/03/02 15:19:18 [DEBUG] ("storeKeyPrefix(apps/frontend/env)") Consul returned 1 key pairs
2016/03/02 15:19:18 [INFO] (view) "storeKeyPrefix(apps/frontend/env)" received data
2016/03/02 15:19:18 [INFO] (runner) running
2016/03/02 15:19:18 [DEBUG] (runner) setting DEBUG="2" from "storeKeyPrefix(apps/frontend/env)"
2016/03/02 15:19:18 [INFO] (runner) missing data for "storeKeyPrefix(app/frontend/supervisor/frontend-render)"
2016/03/02 15:19:18 [DEBUG] ("storeKeyPrefix(app/frontend/supervisor/frontend-render)") Consul returned 0 key pairs
2016/03/02 15:19:18 [INFO] (view) "storeKeyPrefix(app/frontend/supervisor/frontend-render)" received data
2016/03/02 15:19:18 [INFO] (runner) running
2016/03/02 15:19:18 [DEBUG] (runner) setting DEBUG="2" from "storeKeyPrefix(apps/frontend/env)"
2016/03/02 15:19:18 [DEBUG] Environment:
2016/03/02 15:19:18 [DEBUG]   DEBUG="2"
2016/03/02 15:19:18 [INFO] (runner) running command bash -c ./node_modules/.bin/babel-node src/server/server.js
{"name":"frontend-app","hostname":"local","pid":1048,"level":30,"msg":"Bownty frontend app listening at http://:::58230 (PID: 1048)","time":"2016-03-02T15:19:20.325Z","v":0}
{"name":"frontend-app","hostname":"local","pid":1048,"section":"consul-lifecycle","level":40,"msg":"Service f4f498df-f41d-4d04-83ac-2587c4aac45c is _NOT_ healthy","time":"2016-03-02T15:19:20.336Z","v":0}
{"name":"frontend-app","hostname":"local","pid":1048,"section":"lifecycle","level":30,"msg":"Registered consul service check...","time":"2016-03-02T15:19:20.339Z","v":0}
{"name":"frontend-app","hostname":"local","pid":1048,"section":"consul-lifecycle","level":40,"msg":"Unhealthy service f4f498df-f41d-4d04-83ac-2587c4aac45c was de-registered","time":"2016-03-02T15:19:20.348Z","v":0}
Received interrupt, cleaning up...
2016/03/02 15:19:40 [INFO] (runner) stopping
2016/03/02 15:19:40 [INFO] (watcher) stopping all views
2016/03/02 15:19:40 [DEBUG] (watcher) stopping "storeKeyPrefix(apps/frontend/env)"
2016/03/02 15:19:40 [DEBUG] (watcher) stopping "storeKeyPrefix(app/frontend/supervisor/frontend-render)"
2016/03/02 15:19:40 [DEBUG] (runner) killing child process
2016/03/02 15:19:40 [INFO] (runner) waiting 1.95s for random splay
www-data @ local.bownty.net [ /home/cw/www/frontend/htdocs ] [ 15:19:40 ]  (master *=)
->
```

command 3

```
www-data @ local.bownty.net [ /home/cw/www/frontend/htdocs ] [ 15:19:32 ]  (master *=)
-> killall envconsul
```

command 4

```
www-data @ local.bownty.net [ /home/cw/www/frontend/htdocs ] [ 15:19:40 ]  (master *=)
-> ps xau|grep 1048 # command 3
www-data  1048  6.9  2.7 847460 141936 pts/1   Sl   15:19   0:01 /opt/nodejs/node-v5.7.0-linux-x64/bin/node /var/www/frontend/htdocs/node_modules/babel-cli/lib/_babel-node src/server/server.js
```

so the process still exist, even though envconsul claim to have killed it

**Comment by sethvargo** [comment]
Sending SIGTERM to envconsul kills the envconsul process, not the child process itself. Envconsul does not do process reaping unless you specify the `-reap` flag, since the expectation is that you're running under an init system that manages those children.

In your example, you're also running a bash sub-shell for the process, so the signal will be sent to bash, not the process its running. Does that make sense?

**Comment by jippi** [comment]
@sethvargo i can't find any docs on `-reap` ? :)

**Comment by sethvargo** [comment]
Hrm maybe we only added that to Consul Template and not envconsul. Nonetheless, I think you're experiencing this issue because of a bug in Golang 1.5 where childprocesses were lost. That's fixed in 1.6, so a recompile should fix this behavior.

**Comment by jippi** [comment]
@sethvargo when do you plan to cut a new envconsul with go 1.6 ? :)

**Comment by cottrell** [comment]
Has this bee fixed in version 0.6.2? I am seeing ctrl-c be communicated as effectively as ctrl-d in a wrapped python process.
